infrastructure/src/main/java/org/corfudb/infrastructure/management/FullyConnectedClusterRecommendationEngine.java

```diff
@@ -0,0 +1,506 @@
+            clusterMatrix.put(currentNode, unreachableNeighborsSet);
+
+            // Capture symmetric link failures for the neighbors of currentNode
+            unreachableNeighborsSet
+                    .stream()
+                    .forEach(unreachableNeighbor ->
+                            clusterMatrix
+                                    .computeIfAbsent(unreachableNeighbor, key -> new HashSet<>())
+                                    .add(currentNode));
+
+        }
+
+        return clusterMatrix;
+    }
+
+    /**
+     * Create a map of Corfu nodes which is the superset of failed nodes along with the neighbors
+     * for each node. Each of these nodes are currently amongst the active nodes in the provided
+     * {@link Layout} and have at least one link failure with their neighboring Corfu servers.
+     *
+     * @param layout an instance of {@link Layout} representing expected layout capturing amongst
+     *               the  other things, information about responsive and unresponsive servers.
+     * @param nodeFailedNeighborMap A map which each of its entries corresponds to a Corfu
+     *                              server and a set of its neighboring nodes where a link failure
+     *                              has been observed. In other word, each entry represents a
+     *                              collection of edges from the source captured as the key of
+     *                              that entry to its failed neighbors.
+     * @return A map representing Corfu servers considered as responsive in the provided layout and
+     *         that have failed links to at least another responsive servers from the provided
+     *         layout. Mathematically the map represents a symmetric matrix which both columns
+     *         and rows represent active nodes
+     */
+    private Map<String, Set<String>> potentialFailedSuspects(Layout layout,
+                                                             Map<String, Set<String>> nodeFailedNeighborMap) {
+
+        final Map<String, Set<String>> failedLinkNodeNeighborMap = new HashMap<>();
+
+        nodeFailedNeighborMap
+                .entrySet()
+                .stream()
+                .filter(nodeNeighborEntry -> layout.getAllActiveServers()
+                                                   .contains(nodeNeighborEntry.getKey()))
+                .forEach(suspectedFailedNodeEntry -> {
+                    // Remove the failed neighbors that are not part of layout's active servers
+                    final Set<String> neighbors = suspectedFailedNodeEntry.getValue();
+                    neighbors.retainAll(layout.getAllActiveServers());
+                    failedLinkNodeNeighborMap.put(suspectedFailedNodeEntry.getKey(), neighbors);
+                });
+
+        // Return the candidates which their failing neighbors are non-empty set of active servers
+        return failedLinkNodeNeighborMap
```

Comment:
why do you need a second loop here ?

Reply:
The first loop creates a map by collecting the nodes that observe link failure along with active neighbors to which a link failure has been observed (edge failures). The second loop filters those whose responsive neighbors are empty. Could have combined the two but would make it too complicated without much benefits.